### PR TITLE
RSDK-9299 - refactor get_resource_name to be a non-proto method

### PR DESCRIPTION
**Pull request by stuqdog**

### CODE REVIEW

File: src/viam/sdk/common/utils.cpp [hunk 1]
```diff
@@ -183,6 +184,26 @@ bool from_dm_from_extra(const ProtoStruct& extra) {
     }
     return false;
 }
+std::pair<std::string, std::string> long_name_to_remote_and_short(const std::string& long_name) {
+    std::vector<std::string> name_parts;
+    // boost::split causes a clang-tidy false positive, see
+    // https://bugs.llvm.org/show_bug.cgi?id=41141
+    //
+    // NOLINTNEXTLINE
+    boost::split(name_parts, long_name, boost::is_any_of(":"));
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/utils.cpp:193]: good question! I tend to like fully qualified names when they're not overly verbose. I appreciate the explicitness. And, a broader `using` line can add more than we need to scope and create potential for namespace collision.

Also in this case, `boost::split` and `boost::is_any_of` are only used once each in this file so `using` them would actually make the code (slightly) more verbose, not less.

File: src/viam/sdk/common/utils.hpp [hunk 2]
```diff
@@ -61,6 +61,10 @@ class ClientContext {
     grpc::ClientContext wrapped_context_;
 };
 
+/// @brief Given a fully qualified resource name, returns remote name (or "" if no remote name
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/utils.hpp:64]: a remote name is the name of a remote that the resource is part of, and the short name is just the name of the resource. so we could have `remote1:my_arm` as a long name, which is comprised of a remote name (`remote1`) and a short name `my_arm`).

File: src/viam/sdk/common/utils.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/utils.cpp:193]: I don't think this is possible. A long name is of form `{remote}:{short_name}` (or `{remote1}:...{remoteN}:{short_name}`. If there's no colon then we just have a short name so we just return the name as-is, which is ideal.

File: src/viam/sdk/components/component.cpp [hunk 3]
```diff
@@ -2,26 +2,17 @@
...
 
-Component::Component(std::string name) : Resource(std::move(name)){};
+Component::Component(std::string name) : Resource(std::move(name)) {}
 
-ResourceName Component::get_resource_name(std::string name) const {
-    auto r = this->Resource::get_resource_name(name);
-    *r.mutable_type() = kComponent;
-    return r;
+Name Component::get_resource_name() const {
+    return Resource::get_resource_name(kComponent);
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/components/component.cpp:15]: `kComponent` and `kService` are defined in `utils.hpp`, which is included in this file (and in `service.cpp` below).

File: src/viam/sdk/services/service.cpp [hunk 4]
```diff
@@ -2,21 +2,16 @@
...
 
 namespace viam {
 namespace sdk {
 
-common::v1::ResourceName Service::get_resource_name(std::string name) const {
-    auto r = this->Resource::get_resource_name(name);
-    *r.mutable_type() = kService;
-    return r;
+Name Service::get_resource_name() const {
+    return Resource::get_resource_name(kService);
```
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/services/service.cpp:11]: see above

**Comment by stuqdog** [comment]
> left a few questions for my own understanding but other than that looks good to me. are there any specific tests for this? also what is the reasoning for changing get_resource_name to be non-proto?

good questions! no new tests but there are existing proto conversion tests. The reasoning is that proto types affect the ABI but we don't have control their shape, so we want to remove them from the user API as much as possible. Also, a `Name` is basically our custom shadow type that is easier to interact with, so we want to encourage users to use a `Name` instead of a proto `resourcename`.

File: src/viam/sdk/common/utils.cpp [hunk 5]
```diff
@@ -183,6 +184,26 @@ bool from_dm_from_extra(const ProtoStruct& extra) {
...
+    auto name = name_parts.back();
+    name_parts.pop_back();
+    auto remote_name = name_parts.empty()
```
[the remark this answers is not in the document]
Comment (lia-viam) [reply src/viam/sdk/common/utils.cpp:196]: Oh wait just saw your reply to Julie above, so split will return a vector with one element if there's no separator?

[the remark this answers is not in the document]
Comment (stuqdog) [reply src/viam/sdk/common/utils.cpp:196]: > Oh wait just saw your reply to Julie above, so split will return a vector with one element if there's no separator?

Yep! I've just added a test to highlight a couple cases just to be be explicit.

**Review by lia-viam — APPROVED** [review]

**Pull request merged**